File: PathoFroh/src/main/java/com/patho/main/action/dialog/task/CreateTaskDialog.java
package com.patho.main.action.dialog.task;

import com.patho.main.action.dialog.AbstractDialog;
import com.patho.main.action.dialog.task.CreateTaskDialog.TaskTempData.SampleTempData;
import com.patho.main.action.handler.MessageHandler;
import com.patho.main.common.*;
import com.patho.main.model.MaterialPreset;
import com.patho.main.model.PDFContainer;
import com.patho.main.model.StainingPrototype;
import com.patho.main.model.patient.Block;
import com.patho.main.model.patient.Patient;
import com.patho.main.model.patient.Sample;
import com.patho.main.model.patient.Task;
import com.patho.main.model.patient.miscellaneous.BioBank;
import com.patho.main.service.impl.SpringContextBridge;
import com.patho.main.template.DocumentToken;
import com.patho.main.template.PrintDocument;
import com.patho.main.ui.selectors.StainingPrototypeHolder;
import com.patho.main.util.dialog.event.PDFSelectEvent;
import com.patho.main.util.dialog.event.PatientReloadEvent;
import com.patho.main.util.exception.CustomNotUniqueReqest;
import com.patho.main.util.exceptions.TaskNotFoundException;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.helper.TaskUtil;
import com.patho.main.util.pdf.PDFCreator;
import com.patho.main.util.pdf.PrintOrder;
import lombok.Getter;
import lombok.Setter;
import org.primefaces.event.SelectEvent;
import org.springframework.transaction.TransactionStatus;
import org.springframework.transaction.support.TransactionCallback;

import javax.faces.application.FacesMessage;
import javax.faces.component.UIComponent;
import javax.faces.context.FacesContext;
import javax.faces.validator.ValidatorException;
import java.io.FileNotFoundException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Getter
@Setter
public class CreateTaskDialog extends AbstractDialog {

    private Patient patient;

    private List<MaterialPreset> materialList;

    private TaskTempData taskData;

    /**
     * Initializes the bean and shows the createTaskDialog
     *
     * @param patient
     */
    public CreateTaskDialog initAndPrepareBean(Patient patient) {
        if (initBean(patient))
            prepareDialog();

        return this;
    }

    /**
     * Initializes the bean
     *
     * @param patient
     */
    public boolean initBean(Patient patient) {
        super.initBean(null, Dialog.TASK_CREATE, false);

        setPatient(patient);

        // setting material list
        setMaterialList(SpringContextBridge.services().getMaterialPresetRepository().findAllIgnoreArchivedOrderByPriorityCountDesc(true, true));

        setTaskData(new TaskTempData());

        return true;
    }

    /**
     * Creates a new Task object and calls createBiobak at the end.
     *
     * @throws CustomNotUniqueReqest
     */
    public Task createNewTask() {
        return SpringContextBridge.services().getTransactionTemplate().execute(new TransactionCallback<Task>() {

            public Task doInTransaction(TransactionStatus transactionStatus) {

                logger.debug("Creating new Task");

                Task task = SpringContextBridge.services().getTaskService().createTask(getPatient(), getTaskData().getTaskID(), true);
                task.setReceiptDate(getTaskData().getReceiptDate());
                task.setDueDate(getTaskData().getDueDate());
                task.setDateOfSugery(getTaskData().getReceiptDate());
                task.setTaskPriority(getTaskData().getTaskPriority());
                task.setUseAutoNomenclature(getTaskData().isUseAutoNomenclature());

                task = SpringContextBridge.services().getTaskRepository().save(task, resourceBundle.get("log.patient.task.update", task), patient);

                for (SampleTempData sampleTempData : getTaskData().getSamples()) {
                    logger.debug("Creating sample {}", sampleTempData.getMaterial());
                    SpringContextBridge.services().getSampleService().createSample(task, sampleTempData.getMaterialPreset(), sampleTempData.getMaterial(),
                            false, true, true);

                    // creating block manually
                    Sample sample = HistoUtil.getLastElement(task.getSamples());
                    SpringContextBridge.services().getBlockService().createBlock(sample, false, true, false);

                    Block block = HistoUtil.getLastElement(sample.getBlocks());

                    SpringContextBridge.services().getSlideService().createSlidesXTimes(sampleTempData.getStainings(), block, "", "", false, true, false, false);

                }

                // creating standard diagnoses
                task = SpringContextBridge.services().getDiagnosisService().createDiagnosisRevision(task, DiagnosisRevisionType.DIAGNOSIS);

                task = SpringContextBridge.services().getReportIntentService().addReportIntent(task, getPatient().getPerson(), ContactRole.PATIENT, false, false, true).getFirst();

                BioBank bioBank = SpringContextBridge.services().getBioBankService().createBioBank(task);

                task = SpringContextBridge.services().getFavouriteListService().addTaskToList(task, PredefinedFavouriteList.StainingList.getId());

                if (getTaskData().isExternalSamples()) {
                    task = SpringContextBridge.services().getFavouriteListService().addTaskToList(task, getTaskData().getExnternalSampleCommentary(),
                            PredefinedFavouriteList.ReturnSampleList.getId());
                }

                if (getTaskData().getBioBank().isPDFSelected()) {
                    Patient p = SpringContextBridge.services().getPatientRepository().findOptionalById(getPatient().getId()).get();

                    p = SpringContextBridge.services().getPdfService().initializeDataListTree(p);

                    bioBank.setInformedConsentType(getTaskData().getBioBank().getInformedConsentType());

                    SpringContextBridge.services().getPdfService().movePdf(SpringContextBridge.services().getPdfService().getDataListsOfPatient(p), bioBank,
                            getTaskData().getBioBank().getContainer());

                }

                return task;
            }
        });
    }

    public void createAndHide() {
        Task task = createNewTask();
        hideDialog(new PatientReloadEvent(getPatient(), task, true));
    }

    public void createPrintAndHide() {
        task = createNewTask();

        SpringContextBridge.services().getTaskExecutor().execute(new Thread() {
            public void run() {
                logger.debug("Stargin PDF Generation in new Thread");

                Optional<PrintDocument> printDocument = SpringContextBridge.services().getPrintDocumentRepository()
                        .findByID(SpringContextBridge.services().getPathoConfig().getDefaultDocuments().getTaskCreationDocument());

                if (!printDocument.isPresent()) {
                    logger.error("New Task: No TemplateUtil for printing UReport found");
                    MessageHandler.sendGrowlErrorAsResource("growl.error.critical", "growl.print.error.noTemplate");
                    return;
                }

                printDocument.get().initialize(new DocumentToken("task", task),
                        new DocumentToken("patient", patient));

                PDFContainer container = null;
                try {
                    container = new PDFCreator().createPDF(printDocument.get());
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                    MessageHandler.sendGrowlErrorAsResource("growl.error.critical", "growl.print.error.creatingPDF");
                    return;
                }

                SpringContextBridge.services().getCurrentUserHandler().getPrinter().print(new PrintOrder(container, 1, printDocument.get()));

                logger.debug("PDF Generation completed, thread ended");
            }
        });

        hideDialog(new PatientReloadEvent(getPatient(), task, true));
    }

    public void validateTaskID(FacesContext context, UIComponent componentToValidate, Object value)
            throws ValidatorException {

        if (value == null || value.toString().length() != 6) {
            throw new ValidatorException(
                    new FacesMessage(FacesMessage.SEVERITY_ERROR, "", "Auftragsnummer muss sechs Zahlen enthalten."));
        } else if (!value.toString().matches("[0-9]{6}")) {
            throw new ValidatorException(
                    new FacesMessage(FacesMessage.SEVERITY_ERROR, "", "Auftragsnummer darf nur Zahlen enthalten"));
        } else {
            try {
                SpringContextBridge.services().getTaskRepository().findByTaskID(value.toString(), false, false, false, false, false);
                throw new ValidatorException(
                        new FacesMessage(FacesMessage.SEVERITY_ERROR, "", "Auftragsnummer bereits vorhanden"));
            } catch (TaskNotFoundException e) {
            }
        }
    }

    @Getter
    @Setter
    public class TaskTempData {
        private String taskID;
        private LocalDate receiptDate;
        private LocalDate dueDate;
        private boolean useAutoNomenclature;
        private TaskPriority taskPriority;

        private boolean externalSamples;
        private String exnternalSampleCommentary;

        private List<SampleTempData> samples;

        private BioBankTempData bioBank;

        public TaskTempData() {
            this.taskID = SpringContextBridge.services().getTaskService().getNextTaskID();
            this.receiptDate = LocalDate.now();
            this.dueDate = LocalDate.now();
            this.useAutoNomenclature = true;
            this.taskPriority = TaskPriority.NONE;

            this.samples = new ArrayList<SampleTempData>();
            this.samples.add(new SampleTempData(materialList.get(0)));

            this.bioBank = new BioBankTempData();
        }

        public void addSample() {
            samples.add(new SampleTempData(materialList.get(0)));
            updateSampleNames();
        }

        public void removeSample() {
            if (samples.size() > 1)
                samples.remove(samples.size() - 1);

            updateSampleNames();
        }

        public void updateSampleNames() {
            int i = 1;
            for (SampleTempData sampleTempData : samples) {
                sampleTempData.setSampleID(TaskUtil.getSampleName(samples.size(), i, useAutoNomenclature));
                i++;
            }
        }

        @Getter
        @Setter
        public class SampleTempData {

            private MaterialPreset materialPreset;
            private String material;
            private String sampleID;

            private List<StainingPrototypeHolder> stainings = new ArrayList<StainingPrototypeHolder>();
            private boolean showStainings;

            public SampleTempData(MaterialPreset materialPreset) {
                this.materialPreset = materialPreset;
                this.sampleID = "";
                this.showStainings = false;

                onMaterialPresetChange();
            }

            public void onMaterialPresetChange() {
                setMaterial(getMaterialPreset() != null ? getMaterialPreset().getName() : "");

                stainings.clear();

                for (StainingPrototype prototype : getMaterialPreset().getStainingPrototypes()) {
                    System.out.println("adding " + prototype.getName());
                    stainings.add(new StainingPrototypeHolder(prototype));
                }

            }

        }

        @Getter
        @Setter
        public class BioBankTempData {
            private PDFContainer container;
            private InformedConsentType informedConsentType;

            public BioBankTempData() {
                this.container = null;
                this.informedConsentType = InformedConsentType.NONE;
            }

            public boolean isPDFSelected() {
                return container != null;
            }

            public void onMediaSelectReturn(SelectEvent event) {
                if (event.getObject() != null && event.getObject() instanceof PDFSelectEvent) {
                    setContainer(((PDFSelectEvent) event.getObject()).getObj());
                    setInformedConsentType(InformedConsentType.FULL);
                } else {
                    setInformedConsentType(InformedConsentType.NONE);
                }
            }
        }
    }
}

File: PathoFroh/src/main/kotlin/com/patho/main/dialog/task/CreateTaskDialog.kt
package com.patho.main.dialog.task

import com.patho.main.action.handler.CurrentUserHandler
import com.patho.main.action.handler.MessageHandler
import com.patho.main.common.*
import com.patho.main.config.PathoConfig
import com.patho.main.dialog.AbstractDialog_
import com.patho.main.model.MaterialPreset
import com.patho.main.model.PDFContainer
import com.patho.main.model.patient.Patient
import com.patho.main.model.patient.Task
import com.patho.main.repository.jpa.MaterialPresetRepository
import com.patho.main.repository.jpa.PatientRepository
import com.patho.main.repository.jpa.TaskRepository
import com.patho.main.repository.miscellaneous.PrintDocumentRepository
import com.patho.main.service.*
import com.patho.main.service.PDFService.Companion.flatListOfDataLists
import com.patho.main.service.impl.SpringContextBridge
import com.patho.main.service.impl.SpringContextBridge.Companion.services
import com.patho.main.template.DocumentToken
import com.patho.main.ui.selectors.StainingPrototypeHolder
import com.patho.main.util.dialog.event.PDFSelectEvent
import com.patho.main.util.dialog.event.PatientReloadEvent
import com.patho.main.util.exceptions.TaskNotFoundException
import com.patho.main.util.helper.HistoUtil
import com.patho.main.util.helper.TaskUtil
import com.patho.main.util.pdf.LazyPDFReturnHandler
import com.patho.main.util.pdf.PrintOrder
import com.patho.main.util.pdf.creator.PDFCreatorNonBlocking
import org.primefaces.event.SelectEvent
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.transaction.support.TransactionTemplate
import java.time.LocalDate
import javax.faces.application.FacesMessage
import javax.faces.component.UIComponent
import javax.faces.context.FacesContext
import javax.faces.validator.ValidatorException


@Component()
open class CreateTaskDialog @Autowired constructor(
        private val materialPresetRepository: MaterialPresetRepository,
        private val transactionTemplate: TransactionTemplate,
        private val currentUserHandler: CurrentUserHandler,
        private val taskRepository: TaskRepository,
        private val taskService: TaskService,
        private val sampleService: SampleService,
        private val blockService: BlockService,
        private val slideService: SlideService,
        private val diagnosisService: DiagnosisService,
        private val reportIntentService: ReportIntentService,
        private val bioBankService: BioBankService,
        private val favouriteListService: FavouriteListService,
        private val pdfService: PDFService,
        private val patientRepository: PatientRepository,
        private val pathoConfig: PathoConfig,
        private val printDocumentRepository: PrintDocumentRepository) : AbstractDialog_(Dialog.TASK_CREATE) {

    lateinit var patient: Patient

    var materialList: List<MaterialPreset> = listOf()

    lateinit var taskData: TaskTemplate

    open fun initAndPrepareBean(patient: Patient): CreateTaskDialog {
        if (initBean(patient))
            prepareDialog()
        return this
    }

    open fun initBean(patient: Patient): Boolean {
        this.patient = patient

        // setting material list
        materialList = materialPresetRepository.findAllOrderByPriorityCountDesc(true, true)

        taskData = TaskTemplate(materialList)

        return super.initBean()
    }

    /**
     * Creates a new Task and a Biobank object.
     * TODO: Move to service
     */
    private fun createNewTask(): Task? {
        return transactionTemplate.execute {
            logger.debug("Creating new Task")
            var task = taskService.createTask(patient, taskData.taskID, true)
            task.receiptDate = taskData.receiptDate
            task.dueDate = taskData.dueDate
            task.dateOfSugery = taskData.receiptDate
            task.taskPriority = taskData.taskPriority
            task.useAutoNomenclature = taskData.isUseAutoNomenclature
            task = taskRepository.save(task, resourceBundle["log.patient.task.update", task], patient)

            for (sampleTempData in taskData.samples) {
                logger.debug("Creating sample ${sampleTempData.material}")
                sampleService.createSample(task, sampleTempData.materialPreset, sampleTempData.material,
                        false, true, true)
                // creating block manually
                val sample = HistoUtil.getLastElement(task.samples)
                blockService.createBlock(sample, false, true, false)
                val block = HistoUtil.getLastElement(sample.blocks)
                slideService.createSlidesXTimes(sampleTempData.stainings, block, "", "", false, true, false, false)
            }

            // creating standard diagnoses
            task = diagnosisService.createDiagnosisRevision(task, DiagnosisRevisionType.DIAGNOSIS)
            task = reportIntentService.addReportIntent(task, patient.person, ContactRole.PATIENT, false, false, true).first

            val bioBank = bioBankService.createBioBank(task)
            task = favouriteListService.addTaskToList(task, PredefinedFavouriteList.StainingList.id)

            if (taskData.isExternalSamples) {
                task = favouriteListService.addTaskToList(task, taskData.externalSampleCommentary,
                        PredefinedFavouriteList.ReturnSampleList.id)
            }

            if (taskData.bioBank.isPDFSelected) {
                var p = patientRepository.findOptionalById(patient.id).get()
                p = pdfService.initializeDataListTree(p)
                bioBank.informedConsentType = taskData.bioBank.informedConsentType
                if (taskData.bioBank.container != null)
                    pdfService.movePDF(flatListOfDataLists(p)!!, bioBank, taskData.bioBank.container!!)
            }
            task
        }
    }

    open fun createAndHide() {
        val task = createNewTask()
        hideDialog(PatientReloadEvent(patient, task, true))
    }

    open fun createPrintAndHide() {
        val task = createNewTask()
        val currentPrinter = currentUserHandler.printer

        logger.debug("Stargin PDF Generation in new Thread")

        val printDocument = printDocumentRepository.findByID(pathoConfig.defaultDocuments.taskCreationDocument)

        if (!printDocument.isPresent) {
            logger.error("New Task: No TemplateUtil for printing UReport found")
            MessageHandler.sendGrowlErrorAsResource("growl.error.critical", "growl.print.error.noTemplate")
            return
        }

        printDocument.get().initialize(DocumentToken("task", task),
                DocumentToken("patient", patient))

        PDFCreatorNonBlocking().create(printDocument.get(), pathoConfig.fileSettings.workDirectory, false,
                LazyPDFReturnHandler() { container: PDFContainer, s: String ->
                    currentPrinter!!.print(PrintOrder(container, 1, printDocument.get()))
                })

        hideDialog(PatientReloadEvent(patient, task, true))
    }

    open fun validateTaskID(context: FacesContext, componentToValidate: UIComponent, value: Any?) {
        if (value == null || value.toString().length != 6) {
            throw ValidatorException(
                    FacesMessage(FacesMessage.SEVERITY_ERROR, "", resourceBundle["dialog.createTask.validation.error.numberNotMatching"]))
        } else if (!value.toString().matches(Regex("[0-9]{6}"))) {
            throw ValidatorException(
                    FacesMessage(FacesMessage.SEVERITY_ERROR, "", resourceBundle["dialog.createTask.validation.error.onlyNumbers"]))
        } else {
            try {
                services().taskRepository.findByTaskID(value.toString(), false, false, false, false, false)
                throw ValidatorException(
                        FacesMessage(FacesMessage.SEVERITY_ERROR, "", resourceBundle["dialog.createTask.validation.error.doubleNumber"]))
            } catch (e: TaskNotFoundException) {
            }
        }
    }

    class TaskTemplate(private var materialList: List<MaterialPreset>) {
        var taskID: String = SpringContextBridge.services().taskService.getNextTaskID()
        var receiptDate: LocalDate = LocalDate.now()
        var dueDate: LocalDate = LocalDate.now()
        var isUseAutoNomenclature: Boolean = true
        var taskPriority: TaskPriority = TaskPriority.NONE;

        var isExternalSamples = false
        var externalSampleCommentary: String = ""

        val samples: MutableList<SampleTempData> = mutableListOf()

        val bioBank: BioBankTempData = BioBankTempData()

        init {
            samples.add(SampleTempData(materialList.first()))
        }

        fun addSample() {
            samples.add(SampleTempData(materialList.first()))
            updateSampleNames()
        }

        fun removeSample() {
            if (samples.size > 1) samples.removeAt(samples.size - 1)
            updateSampleNames()
        }

        fun updateSampleNames() {
            var i = 1
            for (sampleTempData in samples) {
                sampleTempData.sampleID = TaskUtil.getSampleName(samples.size, i, isUseAutoNomenclature)
                i++
            }
        }

        class SampleTempData(var materialPreset: MaterialPreset?) {
            var sampleID: String = ""
            var material: String = ""
            var showStainings: Boolean = false
            val stainings: MutableList<StainingPrototypeHolder> = mutableListOf()

            fun onMaterialPresetChange() {
                material = materialPreset?.name ?: ""
                stainings.clear()

                materialPreset?.stainingPrototypes?.forEach {
                    stainings.add(StainingPrototypeHolder(it))
                }
            }
        }
    }

    class BioBankTempData {
        var container: PDFContainer? = null
        var informedConsentType: InformedConsentType = InformedConsentType.NONE

        val isPDFSelected
            get() = container != null

        fun onMediaSelectReturn(event: SelectEvent) {
            val obj = event.getObject()

            if (obj != null && obj is PDFSelectEvent) {
                container = (event.getObject() as PDFSelectEvent).obj
                informedConsentType = InformedConsentType.FULL
            } else {
                informedConsentType = InformedConsentType.NONE
            }
        }

    }
}